Hexdecimal: Add example for Unicode code point normative notation

wismill opened the pull request:
`text-builder-linear` tries to be minimalistic. Yet it would be nice to add some examples on how one can efficiently implement special use cases, adapting the code of this package.

The example I took is maybe too complex or too verbose. This PR is about figuring out what examples are useful and if we should expose some lower-level functions (in the present case: `unsafeReplicate`, `lengthAsHex`.

Bodigrim commented on src/Data/Text/Builder/Linear/Buffer.hs:
```diff
@@ -193,3 +198,61 @@ foldlIntoBuffer f = go
+-- >>> :{
+-- (|>&&) :: Buffer %1 -> Char -> Buffer
+-- (|>&&) buf ch =
+--   appendBounded 8 (\dst dstOff -> unsafeAppendCodePoint dst dstOff (ord ch)) buf
```
Why not `appendExact 6`?

Bodigrim replied:
If that's because code points outside of BMP need up to 8 characters, it would be nice to call out this intricacy in the description of the problem.

wismill replied:
```haskell
maxBound @Char
-- '\x10ffff'
length "U+10FFFF"
-- 8
```

wismill replied:
So length would be between 6 (e.g. `U+0000`) and 8 (e.g. `U+10FFFF`).

wismill replied:
Did not get your 2nd message in time. Yes, I should update the description to make this clear. Again, this is just an example for discussion: the idea is to provide examples of writing efficient formatting functions that would not fit in a minimalistic library.

Bodigrim replied:
Yeah, I'm fine with adding such piece of documentation, thanks.

wismill replied:
Done

wismill commented:
Maybe it would be better to move this to tests (in a dedicated file) and add a reference to it in the doc: that way, we can have exhaustive doc about this code without making the  library code bloated.

Actually, we could have a directory `examples` with its own test suite and benchmark.  The benchmark could compare the optimized function vs a function created only with the buffer API.  Then it would be easy to extend with new examples.

Bodigrim commented:
An interesting question is whether one can implement this function with the non-`Core` API at all. Prepending `U+` and padding are fine, but there are no case conversions. 

Despite all intricacies, most of the time one can convert upper/lower case in place, which would be much more efficient than what `text` can offer. I wonder if such functionality is a good investment for this package.

wismill commented:
@Bodigrim I moved the Unicode example to a dedicated `examples` directory. I added test & benchmark. I left a link in the  hexadecimal doc that point to the `Unicode.hs` file in this Github repo.

Bodigrim commented:
Thanks, you are extremely thorough!
